src/sampler.jl

```diff
@@ -53,12 +53,20 @@ function adapt!(
     return h, τ, isadapted
 end
 
-function ProgressMeter.next!(pm, stat::NamedTuple, i::Int, metric)
+"""
+Progress meter update with all trajectory stats, iteration number and metric shown.
+"""
+function pm_next!(pm, stat::NamedTuple, i::Int, metric::AbstractMetric)
```

Comment:
Nice trick to use dispatch-based reporting. Perhaps consider 

- renaming `pm_next!` ==> `show_progress!` for clarity? 
- use keyward arguments, instead of positional arguments for `iteration` and `metric`, e.g.

```julia
simple_pm_next!(pm::NoReport, stat::NamedTuple; kargs...) = ProgressMeter.next!(pm)
simple_pm_next!(pm::Progress, stat::NamedTuple; kargs...) = ProgressMeter.next!(pm, (stat...,kargs...))
```

Reply:
It will be nice if we can replace all `progress` flags with `r::AbstractReport`, and define a void `NoReport` type for the case `progress=false`. 

@cpfiffer  Does this make sense to be part of `Interface.jl`?

Reply:
Looks cool! I guess we have to wait until `Interface.jl` is transferred into `MCMCChains.jl` before we can use this `callback` functionality?